### ISSUE TITLE
[cliquet] Timestamp uniqueness by parent_id under heavy load

### ISSUE THREAD

**Issue by tarekziade** [Kinto/kinto#602]
Original : https://github.com/mozilla-services/cliquet/issues/25

summary: we can improve the timestamp creation for postgres to avoid collisions on very high-load. But the scenario to have that issue is unlikely. (same user hammerting our server)

http://stackoverflow.com/questions/34817760/how-to-generate-unique-timestamps-in-postgresql

**Comment by leplatrem** [comment]
The only way to get of this is to rely on a SEQUENCE to build timestamps. I suggest that we get rid of the fact that timestamps are based on clock and make them fully opaque (hence the label protocol:v2)

**Comment by Natim** [comment]
This is something that we definitely want to fix for the kinto webextension stack.

Refs: https://sentry.prod.mozaws.net/operations/kinto-webextensions-stage/issues/384592/

``Key (id, parent_id, collection_id)=(81d2f645-5b5f-2e53-fc0c-913e938ef2a4, , bucket) already exists.``

**Comment by leplatrem** [comment]
The `records_pkey` error that we can see in the mentioned Sentry error is indeed easy to catch, since it's a conflict on `(id, parent_id, collection_id)`, and not on timestamp unicity (`idx_records_parent_id_collection_id_last_modified`).

However, how should we react to it? If it is the same user on both ends, then do nothing or overwrite is fine. But if it's two different users, then the second one should probably receive a unicity error.

**Comment by glasserc** [comment]
So this is two different constraint violations, right?

**Comment by leplatrem** [comment]
Yes, there are mainly two scenarios:
* two overlapping requests trying to create the same object (`records_pkey`)
* two overlapping requests (*same millisec*) creating/updating/deleting objects within the same parent (`idx_records_parent_id_collection_id_last_modified`)

The constraint error is not raised when the operation is performed but when the transaction is committed — just before serving the response.

The approach we took in #1120 does not seem right. Not only that it only fixes the first one, but also mainly because it tries to catch and handle an error that is necessary to maintain a consistent DB state. For example, a rejected insertion in storage rollbacks the changes that were made in permissions. If we swallow it, what do we do with permissions? And why would we merge the attributes of the rejected object into the one that was concurrently committed?

For the second constraint, one approach (as suggested[ on SO](http://stackoverflow.com/questions/34817760/how-to-generate-unique-timestamps-in-postgresql)) is to rely on PG sequences. But that breaks our protocol since timestamps would not be epoch anymore. 

Currently, I don't see any obvious way to magically handle these situations and prevent error responses. IMO one of the concurrent requests has to be accepted, and the other ones rejected as conflict. 

So maybe, we should simply just rework our error response semantics. When a database error occurs, we return a `503 Service Unavailable` — why not. But what if we would return a `409 Conflict` when an integrity constraint occurs? In the end, that's exactly what is happening. And QA wouldn't have the feeling of breaking the service with concurrent requests...

**Comment by Natim** [comment]
I agree with your analyse. I guess we are now ok to move out of sequences at some point. It is possible to handle timestamps differently.

If think for the default bucket we shouldn't raise a 409 when a conflict on automatic bucket and automatic collection creation occurs because it should be transparent.

Otherwise I agree with what you said.

**Comment by Natim** [comment]
#1122 Add an ignore_conflict argument to create for the default_bucket
